### PR TITLE
Add to cart #9

### PR DESCRIPTION
**Pull request by tmikeschu**
I still need to complete the test for the Cart poro, but the functionality is there. Pull down this branch if you want to work on styling for any cart stuff. 

I will finish up the test and get the major hound violations out.

### CODE REVIEW

**Comment by tmikeschu** [comment]
Tests are all set and add to cart functionality is complete.

I added an optional quantity parameter for the .add_item method in the Cart PORO. Thoughts? The user story doesn't call for it, and if we lean to that, we should take quantity out of the views.

**Comment by tmikeschu** [comment]
@kheppenstall @wlffann @akintner : ready for merge when you are

File: app/models/cart.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  def items
+    item_ids = contents.keys
+    Item.where(id: item_ids)
+    # contents.keys.map { |item_id| Item.find(item_id) }
+  end
+
+  def total_price
+    items.sum(:price_per_unit)
+  end
+end
```
Comment (kheppenstall) [inline app/models/cart.rb]: I like how you incorporated the quantity here. 👍

File: app/models/cart.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  def count_of(item_id)
+    contents[item_id.to_s]
+  end
+
+  def items
+    item_ids = contents.keys
+    Item.where(id: item_ids)
+  end
+
+  def total_price
```
Comment (kheppenstall) [inline app/models/cart.rb]: @tmikeschu 

When calculating total_price should we also be using the quantity in the contents hash? I may be misunderstanding this but it looks like items is only counting unique items. If I'm not off my rocker would something like this work?

`contents.keys.reduce(0) {|sum, id| sum += Item.where(id).total_price * contents[id] }`

or

`items.reduce(0) {|sum, item| sum += item.total_price * contents[item.id] }`

**Comment by wlffann** [comment]
Looks good to me! I like all those PORO methods.

File: app/controllers/carts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  include ActionView::Helpers::TextHelper
+
+  def create
+    item = Item.find(params[:item_id])
+
+    @cart.add_item(item.id)
+    session[:cart] = @cart.contents
+
+    flash[:notice] = "You now have #{pluralize(@cart.count_of(item.id), item.name)}."
+    redirect_back(fallback_location: items_path)
```
Comment (kheppenstall) [inline app/controllers/carts_controller.rb]: Love the redirect_back. Never seen it before.

File: spec/features/carts/visitor_adds_item_to_cart_spec.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      expect(page).to have_content "Total Price: #{@item1.price_per_unit}"
+    end
+
+    scenario "adds an item to their cart from a category page" do
+      category = create(:category_with_items)
+      item = category.items.first
+
+      visit category_path(category.title)
+
+      click_button("Add to Cart", match: :first)
```
Comment (kheppenstall) [inline spec/features/carts/visitor_adds_item_to_cart_spec.rb]: `match: :first` is a nice trick.

File: app/controllers/carts_controller.rb (hunk 3 above)
Comment (tmikeschu) [inline app/controllers/carts_controller.rb]: New feature! I was using redirect_to :back, but got a nice warning saying it's deprecated for the next version

**Comment by kheppenstall** [comment]
Lovely work. Once the `total_price` cart method is changed I'm ready for a merger.

File: app/models/cart.rb (hunk 2 above)
Comment (tmikeschu) [inline app/models/cart.rb]: Hole smokes! Can't believe I missed that. You're spot on. 

How about: 
`items.map { |item| item.price_per_unit * contents[item.id.to_s] }.sum`

Comment (tmikeschu) [inline app/models/cart.rb]: (A little more use of Rails and AR methods)

Comment (kheppenstall) [inline app/models/cart.rb]: @tmikeschu 👍 That rails sum method is really useful.

**Comment by akintner** [comment]
@tmikeschu There was a merge conflict with master because of the categories view that just got changed, but it was super small, so I think it's already fixed from here. Checkout your local and see if it's resolved there, too, and then we'll finalize this PR and merge.

**Comment by tmikeschu** [comment]
@akintner @kheppenstall @wlffann : I resolved all the conflicts and all tests are passing. Ready to merge. Thanks for the caution!

**Pull request merged**